docs/library/machine.CAN.rst

```diff
@@ -0,0 +1,364 @@
+.. currentmodule:: machine
+.. _machine.CAN:
+
+class CAN -- Controller Area Network protocol
+=============================================
+
+.. warning:: Currently no MicroPython ports have ``machine.CAN`` controllers.
+             This is a design document only.
+
+CAN is a two-wire serial protocol used for reliable real-time message delivery
+between one or more nodes connected to a common bus. CAN 2.0 was standardised in
+ISO-11898, and is now also known as CAN Classic.
+
+There is also a newer, backwards compatible, protocol named CAN FD (CAN with
+Flexible Data-Rate).
+
+CAN support requires a controller (often an internal microcontroller
+peripheral), and an external transceiver to level-shift the signals onto the CAN
+bus.
+
+The ``machine.CAN`` interface is a *low level basic* CAN messaging interface
+that abstracts a CAN controller as an outgoing priority queue for sending
+messages, an incoming queue for receiving messages, and mechanisms for reporting
+errors.
+
+.. note:: The forthcoming ``can`` and ``aiocan`` micropython-lib modules are the
+          recommended way to use CAN with MicroPython.
+
+Constructor
+-----------
+
+.. class:: CAN(id, **kwargs)
+
+   Construct a CAN controller object of the given id:
+
+    - ``id`` identifies a particular CAN controller object; it is board and port specific.
+    - ``**kwargs`` are hardware-specific identifiers for the particular
+      board and port. For example, to set which GPIO pins to use for the
+      CAN controller.
+
+Methods
+-------
+
+.. method:: CAN.init(bitrate, mode=CAN.Mode.NORMAL, sample_point=75, sjw=None, tseg1=None, tseg2=None, f_clock=None)
+
+   Initialise the CAN bus with the given parameters:
+
+   - ``bitrate`` is the desired bus bit rate in bits per second.
+   - ``mode`` is one of :class:`CAN.Mode` enumerated values, indicating the
+     desired mode of operation.
+   - ``sample_point`` is a percentage of the data bit time. It specifies the
+     position of the bit sample with respect to the whole nominal bit time.
+
+   The remaining parameters all relate to CAN bit timings. If not specified, the CAN controller
+   will pick nominal values based on the specified bit rate.
+
+   Otherwise, they can be specified:
+
+    - ``sjw`` is the resynchronisation jump width in units of time quanta for nominal bits;
```

Comment:
I'd remove all this stuff. It's really hard to use except for the STM32. Even then it's hard. The sample_point value gives you enough control as it adjusts tseg1/tseg2 given f_clock. sjw is a valid setting to provide control for as it's independent of the clock freq. But, only certain ports will be able to set it. On the imxrt it can't be set independently of the tseg1/tseg2.

Reply:
Makes sense. I'll look more closely at some of the other controller implementations and see what's the best cross-compatible way to slice it (I see @kdschlosser made a similar recommendation in their comment). Can always have port-specific optional arguments where available.

I feel like >98% of Microython users will only ever set bitrate anyhow.

Reply:
Sample point control is important however. This is needed to get CAN devices working on buses when you have "un-ideal" bus setups with long runs coming off a bus.

Reply:
Agree. How about slicing it like this?

* Controller-agnostic arguments for `bitrate` and optional custom sample point.
* Controller-specific arguments to explicitly set more if/when needed, with whatever features that controller exposes.

Reply:
This is what I am talking about as far as doing the calculations for a bitrate

https://github.com/kdschlosser/can_timing/blob/main/can_timing/__init__.py

That is an earlier version and I have a more updated version that has more interfaces added to it. But that is the basic idea. standard values fed in and out comes interface specific registers. The user should not need to know how to calculate those registers when all they want to do is connect to a network at a given bitrate.

The other thing that this kind of thing can be used for is connecting to a network without even providing a bitrate. You would use this with the most commonly used bitrates and how it works is the function gets fer in the bitrate returns the registers, the registers get set and then you look for any errors. if there are any errors then there is a pretty good chance that the bitrate being used is not what the network is running at, then move onto the next one. Once valid frames are received and there are no errors then you know you have gotten the correct bitrate.

Reply:
> I'd remove all this stuff. It's really hard to use except for the STM32. 

Looking at these arguments (`sjw`, `tseg1`, `tseg2`) again and comparing the various controller data sheets, I think I'd like to keep them for all controllers. You're right @kwagyeman that most folks won't use them, but they will be useful for fine-tuning parameters in cases like the ones that @kdschlosser is working with.

There seems to be a split in terms of the controller hardware that is configuring using `tseg1 = prop_seg + phase_seg1` (includes: STM32, ESP32, MCP2518FD), and controller hardware that sets `prop_seg` and `phase_seg1` individually (includes: iMXRT, MCP2515).

As per the standard, it's sufficient for a controller to only program the total `tseg1`. So the question becomes which method is more general, and secondarily which method uses less resources.

It's not a clear win either way, but I've kept `tseg1` and `tseg2` mostly to save a parameter. I think drivers for hardware that needs to split this into `prop_seg` and `phase_seg1` should be able to do so pretty easily within the constraints of the controller registers. @kwagyeman does that seem OK to you?

The PR has been updated with some more explanation about this, also.

Reply:
> The only way to keep the API identical across the board for the constructors is going to be using code like this. There is no choice in that. The issue is the interfaces having different registers that need to be set. 

@kdschlosser Just to check, does "code like this" mean code like what's currently proposed in this PR, or code like in the linked `can_timing` project?

Reply:
Everything I am talking about is written in C code, nothing is written in Python. The example I supplied was written to run in MicroPython. It's not pretty but it works. it's also a resource hog and that wouldn't be the case if it was written in C code. The amount of space it takes up in terms of program size is going to be really small.

Reply:
@projectgus 

code like this means the example in that link I posted. code that is able to take simple understoood easy to explain values and turn them into the timing registers needed for an interface.

Reply:
I think it's best to leave the bit rate calculations up to the HAL APIs supplied by each MCU. You only need to control the sample point, bitrate, and sjw to specify everything. As for a tolerance parameter... I think you're goal would always be to hit a specified bitrate and sample point exactly. If you aren't then that would be an error. It's not like people really need the maximum amount of flexibility here.

Anyway, for my purposes I will update the MIMXRT CAN driver to work with whatever is proposed.

Reply:
Thanks @kdschlosser and @kwagyeman for the extra explanation and input. @kdschlosser , I think I understand your reasoning and concerns now. I think we can meet these requirements, although probably with a slightly different design.

> You only need to control the sample point, bitrate, and sjw to specify everything.

I think not quite, though? SJW units are typically time quanta, and this isn't enough information for the caller to know the time quantum length. So would either need to specify SJW as a proportion, or pass additional information here. Let me know if I'm missing a trick here.

---

To try and summarise, I counts five different bit timing methods that have been mentioned somewhere in this PR:

1. Set a nominal `bitrate`, driver chooses the rest. *I'm certain this is all >95% of MicroPython users will ever need, and MP should ensure it is the simplest thing to do.*
2. Set a `bitrate` and a sample point percentage. This allows the driver to calculate everything except for SJW, which will have to be inferred. As @kentindell mentions below, it's easy to get this value wrong.
3. Set `bitrate`, `sjw`, `tseg1` and `tseg2` in units of time quanta. This maps to the CAN ISO standard parameters and (I think) fully defines the bit timings.
4. Set `bitrate`, `sjw`, `prop_seg`, `phase_seg1` and `phase_seg2`. This maps even more closely to the CAN ISO standard, but it's functionally equivalent to 3 (where `tseg1 = prop_seg + phase_seg1`), and the standard even says as much.
5. Provide "first principles" bus physical properties and allowed tolerances, as described above by @kdschlosser. There's an appealing elegance to this, but it's a different approach to what the standard defines which is what most folks working with CAN are more likely to be used to (especially if copying parameters from an existing project, etc.)

At this stage I'm thinking all low-level controller drivers should supports methods 1, 2 & 3. MicroPython can include a generic helper function to calculate method (3) given the inputs from methods 1 & 2 and driver-specific constraints. Method 5 should be possible to implement in Python as a set of functions on top of method 3, and could be bundled in micropython-lib as an addon package.

Reply:
Works for me.

Reply:
The standard defines using the bus length and the node delay as the way to calculate what the SJW is. Most people will just plug a number in and see how it works because after all there is typically only going to be an SJW of 1 to 4. Trial and error in this case is really simple to do. So no big deal there. Those are things that don't need to be added, it can be done is what I was getting at.

The only things that a user would need to supply for the calculations to do their thing would be the bitrate and the SJW. Those 2 things and they would be off and running. They don't need to know a single thing about the CAN specification or how to calculate tseg or propsegs or the timing registers. All of that is done internally as it should be.  Most people that are working with MCU's have an idea of what the bitrate is and the SJW is really only going to come into play with longer busses or buses that have a nigher node count. For the most part that number will end up being set to a 1. The SJW should be defaulted to 1 making the only thing that the user needs to supply is the bitrate. That's it.

I had also written code that does not iterate over the BRP, TSEG1 and TSEG2 ranges. It uses raw calculations instead. The issue there is it is not going to allow any deviation. by doing the iteration method we can use a close bitrate and it would work because of the ability for can to adjust for things like oscillator float. a lot of times you can go as much as +- 5% of the target bitrate. 

If I asked you to tell me what the timing registers are or what the brp, tseg1 and tseg values are for an ESP32V3 running at 80kbps with a sample point of 87.5% could you do it?. Could you locate a table somewhere in the ESP's manual or in the SDK's manual that would tell you those numbers? I already know the answer which is you are not going to be able to locate those numbers. You are not going to find a calculator that would be able to generate the numbers either.

Reply:
If a person has the knowledge of how to get the tseg values and the brp value then they will have the knowledge or at least know how to find how to convert those values into the timing register numbers. This is the reason why I said the only thing that needs to be supplied in terms of an "advanced user" way of doing this is to allow the user to pass the timing registers as a list or tuple to a keyworded parameter "registers" The interface being used is going to know how the length of that passed in tuple or list so it would be able to handle collecting the registers from it without any issues. The list or tuple also allows the common front facing API to be blind as to what is actually being passed. some interfaces only have a single register  some have 2 or 3. The common facing API doesn't need to know what is being passed. That would get handled by the board specific portion.

Reply:
@kdschlosser I've read these latest comments several times and I think I understand them individually, but I honestly don't fully understand what specifically you're asking to change compared to [what's already outlined above](https://github.com/micropython/micropython/pull/13149#discussion_r1448089343) and in the PR.

I did note one suggestion, which was to support initialising with only two parameters: `bitrate` and `sjw`. From your comment I see that "start from sjw=1 and increase until the errors go away" is a reasonable method for finding an SJW value in some use cases, so supporting this seems useful and viable. Thanks, will change.

Regarding exposing hardware-specific register values as an initialisation method, I agree with the comment you made earlier about this being "a support headache". It's reasonable to expose hardware-specific features in initialisation, where they don't have any generalised equivalent, but the driver should abstract this into a common interface as much as possible (similar to the other MicroPython machine drivers). The user who wants to write registers directly always has `machine.mem32` at their disposal.

If I missed or misunderstood an additional API change that you'd like to see, compared to what's already there, then please outline the specific proposed change, and the associated benefit compared to the status quo, as concisely as possible.